Sources/NIOCore/AsyncChannel/AsyncChannel.swift

```diff
@@ -301,6 +302,53 @@ public struct NIOAsyncChannel<Inbound: Sendable, Outbound: Sendable>: Sendable {
             }
         }
 
+        self._outbound.finish()
+        // We ignore errors from close, since all we care about is that the channel has been closed
+        // at this point.
+        self.channel.close(promise: nil)
+        // `closeFuture` should never be failed, so we could ignore the error. However, do an
+        // assertionFailure to guide bad Channel implementations that are incorrectly failing this
+        // future to stop failing it.
+        do {
+            try await self.channel.closeFuture.get()
+        } catch {
+            assertionFailure(
+                """
+                The channel's closeFuture should never be failed, but it was failed with error: \(error).
+                This is an error in the channel's implementation.
+                Refer to `Channel/closeFuture`'s documentation for more information.
+                """
+            )
+        }
+        return result
+    }
+
+    #if compiler(>=6.0)
+    /// Provides scoped access to the inbound and outbound side of the underlying ``Channel``.
+    ///
+    /// - Important: After this method returned the underlying ``Channel`` will be closed.
+    ///
+    /// - Parameters:
+    ///     - actor: actor where this function should be isolated to
+    ///     - body: A closure that gets scoped access to the inbound and outbound.
+    public func executeThenClose<Result>(
+        isolatedTo actor: isolated (any Actor)? = #isolation,
```

Comment:
I think this follows more what we have been doing elsewhere https://github.com/swift-server/swift-service-lifecycle/blob/e2b442cc73b8100f16df23eb2431aa0d4660fc19/Sources/ServiceLifecycle/GracefulShutdown.swift#L41
```suggestion
        isolation actor: isolated (any Actor)? = #isolation,
```